### DISCUSSION TITLE
SSL with custom provider

### DISCUSSION THREAD

**Discussion by afshinpir** [openssl/openssl#24411] [Q&A]
Hi All
Lets assume that we have a provider that fully supports RSA signature which means signature and key management is implemented. But this is the only supported operation from the provider.
Now if we want to create an TLS connection using our provider, how we can achieve it? From this, I'm interested to do TLS's sign/verify operation with provider and the rest with default provider. It is somehow tricky because whole TLS context uses one querystring which is passed in the `SSL_CTX_new_ex`.
I have tried to create SSL  context with null as query string and the pass my EVP_PKEY with my provider query string. But I somehow cannot reliably make it work.
As a side note, I should say that my provider does **not** export private keys.
Any idea what I can do?

**Comment by paulidale** [comment]
Try using `?provider=your-provider` as the query.  This will favour your provider over default but will drop back top default for things you don't implement.

**Comment by afshinpir** [comment]
I could not get it work with `?provider=my-provider` with my `EC` key support. Because when key exchange happens, SSL context will try to create ECDHE with a context from my provider and it does not work properly. Or at least I could not get it work correctly in this method.
But I was successful with another way. I used `?provider=default` as query string of my SSL context and then passed my SSL private key which was created with `provider=my-provider` query string. Since `?provider=default` allows fallback on my provider and my private key is not exportable, for signing phase, context of key which is `provider=my-provider` will be used and everything works.
This was the only way that I could get everything work in my case.

**Comment by t8m** [comment]
> But I was successful with another way. I used `?provider=default` as query string of my SSL context and then passed my SSL private key which was created with `provider=my-provider` query string. Since `?provider=default` allows fallback on my provider and my private key is not exportable, for signing phase, context of key which is `provider=my-provider` will be used and everything works.
> This was the only way that I could get everything work in my case.

Yes, and that is the correct way how to do it. I'd suggest making the `?provider=default` a default query either via the config file or via `EVP_set_default_properties()`. And then using the `provider=my-provider` when creating the private key. Explicit query will override the same properties specified in the default query.
